drawio-image/figure directives to replace drawio

brechtm opened the pull request:
- inherits from the image/figure directives so that draw.io images
  behave the same as images/figures
- this means that all builders will be able to handle draw.io images
  without any custom code

Refer to the README for an overview of the changes compared to the old
drawio directive. See also #28.

Diagrams are now also exported to the .drawio directory beside the
document. Each diagram is exported to a separate subdirectory in this directory named by the hash of the diagram filename and export
options. The base name of the exported image file is the same as that
of the draw.io file. This has these advantages:
- we don't interfere with the image/figure directive copying of image
  files to the output directory
- a diagram needs to be exported only once when two builders use the
  same export format
- more descriptive image filenames for the HTML builder

brechtm commented:
Two tests fail when the whole test suit is run at once (like on CircleCI) for some reason. I'll investigate that.

Modelmat commented on sphinxcontrib/drawio.py:
```diff
@@ -288,20 +268,34 @@ def on_build_finished(app: Sphinx, exc: Exception) -> None:
 
 
+FALLBACK_EXPORT_FORMAT = "png"
+DEFAULT_BUILDER_EXPORT_FORMAT = dict(html="svg", latex="pdf", rinoh="pdf")
```
I'd prefer if this was a normal dictionary expression, i.e.:
```suggestion
DEFAULT_BUILDER_EXPORT_FORMAT = {
	"html": "svg",
	"latex": "pdf",
	"rinoh": "pdf",
}
```

brechtm replied:
Done.

Modelmat commented on sphinxcontrib/drawio.py:
```diff
@@ -65,88 +64,123 @@ def boolean_spec(argument: Any) -> bool:
-        binary_path = self.builder.config.drawio_binary_path
+    export_dir = Path(builder.srcdir) / '.drawio' / sha_key
+    ensuredir(export_dir)
+    filename = Path(input_stem).with_suffix('.' + output_format)
```
I'd prefer this not to just be `folder-1.png`. If possible:
* use the filename of the draw.io file; as the tests have it'd be `box`.
* if the options are changed from the config **defaults**, append them in alphabetical order.

e.g. for a file named box with the zoom changed from 100 to 150:
`box-scale150.png`

brechtm replied:
I'm not sure I'm following you. `box.drawio` will be exported to `.drawio/<hash A>/box.png`. The HTML builder will copy that file to `_build/_html/_images/box.png`. If`box.drawio` is referenced in another `drawio-image` directive with different export options, it will be exported to ``.drawio/<hash B>/box.png`. The HTML builder copies that file to `_build/_html/_images/box1.png`.

I feel that adding the export options to the filename:
- is not so useful since in most cases you will only reference a diagram once,
- makes image filenames unnecessarily long, and
- outside the scope of this PR

Modelmat replied:
Okay then.
This is behaviour seems to be that of the image figure backend, yeah?

brechtm replied:
Which behavior you are referring to? If you're talking about the copying of the image files, that is how the HTML builder handles image nodes.

Modelmat replied:
Yeah, that's what I was trying to get at. That's fine then.

Modelmat commented on README.md:
```diff
@@ -1,44 +1,58 @@
+accept the path to a `.drawio` file and additional options to control
+exporting of the diagram to a format suitable for the selected Sphinx builder. 
+
+These directives replace the `drawio` directive which is now deprecated, and
```
Please add a note that these will be removed in version 1.0 [I think this makes sense?]

brechtm replied:
Done.

Modelmat reviewed (changes requested):
A few requests.

If possible, could you make git recognise the **move** of `drawio.py` to `old_drawio` as a move, and the creation of a new `drawio.py`? Should be able to use `git mv` or somewhat (might need a rebase, and _possibly_ multiple comments but I don't think so).

I'd also prefer if the old file is `old_drawio` not `olddrawio`.

brechtm commented:
Thanks for the quick review!

> If possible, could you make git recognise the **move** of `drawio.py` to `old_drawio` as a move, and the creation of a new `drawio.py`? Should be able to use `git mv` or somewhat (might need a rebase, and _possibly_ multiple comments but I don't think so).

I'm afraid not. `git mv` doesn't do anything special; no "move" or "copy" information is stored in commits, AFAIK. Git identifies file copies after the fact, for example when you do `git log -p`.

> I'd also prefer if the old file is `old_drawio` not `olddrawio`.

I'll change that. I also considered creating a `drawio` package with `__init__.py` (new code) and `deprecated.py` (old code). Let me know if you prefer that approach.

brechtm commented on requirements.txt:
```diff
@@ -1,3 +1,3 @@
-sphinx==2.2.0
+sphinx==3.2.1
```
As it is now, we're not testing against the current Sphinx version at all. Ideally, we'd test against a couple of Sphinx versions. Since implementing that is a sizeable project on its own, I figured we should preferably test against the current version. We can drop the `==3.2.1` to always use the latest version.

Modelmat replied:
Doing `sphinx>=2` would install the latest version I'm pretty sure, so it would work.

brechtm commented on sphinxcontrib/drawio.py:
```diff
@@ -65,88 +64,123 @@ def boolean_spec(argument: Any) -> bool:
-class DrawIONode(nodes.General, nodes.Element):
-    pass
+class DrawIOWarning(Exception):
+    def __init__(self, message):
```
This class is no longer needed. I'm removing the _get_filename_ function I introduced since it's only called in a single place now.

brechtm commented on sphinxcontrib/drawio.py:
```diff
@@ -65,88 +64,123 @@ def boolean_spec(argument: Any) -> bool:
-        "scale": directives.positive_int,
-        "transparency": boolean_spec,
-        "width": directives.positive_int,
+        'format': format_spec,
```
Changed. Old habits die hard :-)
We should look into using [black](https://github.com/psf/black).

Modelmat replied:
Yeah, I'm pretty sure I had it as an issue to add it. Apparently not => #37

brechtm commented on sphinxcontrib/drawio.py:
```diff
@@ -158,10 +192,10 @@ def render_drawio(self: SphinxTranslator, node: DrawIONode, in_filename: str,
-        if option in node["config"]:
-            value = node["config"][option]
-            extra_args.append("--{}".format(option))
+    for option in OPTIONAL_UNIQUES:
```
Done.

brechtm commented on sphinxcontrib/drawio.py:
```diff
@@ -65,88 +64,123 @@ def boolean_spec(argument: Any) -> bool:
+        return image
+
+
+OPTIONAL_UNIQUES = {"export-height": "height",
```
Done.

brechtm commented on sphinxcontrib/drawio.py:
```diff
@@ -210,58 +242,6 @@ def render_drawio(self: SphinxTranslator, node: DrawIONode, in_filename: str,
                                                   exc.stderr, exc.stdout))
 
 
-def render_drawio_html(self: HTMLTranslator, node: DrawIONode) -> None:
```
Correct. As far as the builders are concerned there are no new nodes.

brechtm commented on sphinxcontrib/drawio.py:
```diff
@@ -65,88 +64,123 @@ def boolean_spec(argument: Any) -> bool:
+
+    @staticmethod
+    def _get_image_node(nodes):
+        image, = nodes                   # TODO: test
```
Replaced with more intelligent code.

Modelmat commented on sphinxcontrib/drawio.py:
```diff
@@ -105,7 +110,10 @@ def run(self) -> List[Node]:
         nodes = super().run()
-        image = self._get_image_node(nodes)
+        for node in traverse(nodes):
+            if type(node) == docutils_image:
```
```suggestion
            if isinstance(node, docutils_image):
```

Modelmat commented:
> I'm afraid not. git mv doesn't do anything special; no "move" or "copy" information is stored in commits, AFAIK. Git identifies file copies after the fact, for example when you do git log -p.

Okay.

> I'll change that. I also considered creating a drawio package with __init__.py (new code) and deprecated.py (old code). Let me know if you prefer that approach.

The current one is fine.

---
I'd thought I'd already responded to this, I probably forgot to press send ;)

Modelmat commented:
If you'd also like to rebase your changes onto latest master, that'd be great :)

brechtm commented:
I rebased and renamed _olddrawio.py_.

Thanks for the push access! How about I commit bugfixes myself but still open a PR for feature additions so they are first reviewed?

Modelmat commented:
I think it's probably best to keep everything to PRs, however it probably depends on the nature of the fix - a one line change vs refactoring some method would make a difference to whether it's okay (the first would be fine, the latter, probably not).
That being said, if I take 2 weeks to respond to an bugfix PR you can probably just merge it :)

I'll merge this PR when I'm next at a desktop, and will probably tag a new version on pypi.

brechtm commented:
Just a friendly reminder to merge this and create a new release. We're anxious to start using it :-)

Modelmat commented:
Sorry, my bad. I'm in the middle of exams right now.

If you'd like to merge & tag a new release go for it.

[Not sure if documented but I tend to use lightweight tags on the latest commit then push both tags & commit at the same time].

Modelmat commented:
*Just to be clear, I'm saying that you _should_ merge and tag this because I won't get to it for a while.

brechtm commented:
> Sorry, my bad. I'm in the middle of exams right now.

No problem. Good luck!

Merged, tagged, pushed: 0.0.10 is available on PyPI! We'll start using this now and I'll be the fixing bug we run into. Cheers!
